numba/targets/arraymath.py

```diff
@@ -668,6 +668,101 @@ def _trivial_isnan(x):
         return _trivial_isnan
 
 
+@overload(np.iscomplex)
+def np_iscomplex(x):
+    def impl(x):
+        ax = np.asarray(x) # NumPy uses asanyarray here!
+        return ax.imag != 0
+    return impl
+
+
+@overload(np.isreal)
+def np_isreal(x):
+    def impl(x):
+        ax = np.iscomplex(x)
+        return np.invert(ax)
+    return impl
+
+
+@overload(np.iscomplexobj)
+def iscomplexobj(x):
+    # Implementation based on NumPy
+    # https://github.com/numpy/numpy/blob/d9b1e32cb8ef90d6b4a47853241db2a28146a57d/numpy/lib/type_check.py#L282-L320
+    dt = determine_dtype(x)
+    iscmplx = np.issubdtype(dt, np.complexfloating)
+    return lambda x: iscmplx
+
+
+@overload(np.isrealobj)
+def isrealobj(x):
+    # Return True if x is a not complex type or an array of complex numbers.
+    # Implementation based on NumPy
+    # https://github.com/numpy/numpy/blob/ccfbcc1cd9a4035a467f2e982a565ab27de25b6b/numpy/lib/type_check.py#L290-L322
+    def impl(x):
+        return not np.iscomplexobj(x)
+    return impl
+
+
+@overload(np.isneginf)
+def isneginf(x, out=None):
+
+    if not type_can_asarray(x):
+        raise TypingError("First argument must be array-like")
+
+    if numpy_version <= (1, 12):
+        wrapper = register_jitable(lambda x: np.asarray(x))
+    else:
+        wrapper = register_jitable(lambda x: x)
+
+    if is_nonelike(out):
+        if isinstance(x, (types.Array, types.Sequence, types.SliceType)):
+            def impl(x, out=None):
+                x = np.asarray(x)
+                out = np.zeros(x.shape, dtype=types.boolean)
+                np.logical_and(np.isinf(x), np.signbit(x), out)
+                return wrapper(out)
+        else:
+            def impl(x, out=None):
+                out = np.isinf(x) and np.signbit(x)
+                return wrapper(out)
+    else:
+        def impl(x, out=None):
+            np.logical_and(np.isinf(x), np.signbit(x), out)
+            return wrapper(out)
+
+    return impl
+
+
+@overload(np.isposinf)
```

Comment:
Pretty sure we can just use code generator here to avoid having two implementations that differ by `~`? If it's not clear how to do this then please shout!

Reply:
Can you give me an example?

Reply:
Hi Stuart, can you check my approach if this is what you're expecting.

Reply:
I've taken a look, this approach is similar to what I was thinking, thanks.